File: components/hormann_hcp1/button.py
import esphome.codegen as cg
import esphome.config_validation as cv
from esphome.components import button
from esphome.const import CONF_ID

from . import hormann_hcp1_ns, HormannHCP1Component, CONF_HORMANN_HCP1_ID

DEPENDENCIES = ['hormann_hcp1']

CONF_ACTION = 'action'

HormannButton = hormann_hcp1_ns.class_('HormannButton', button.Button, cg.Component)

ACTION_OPTIONS = {
    'impulse': 'ACTION_IMPULSE',
    'venting': 'ACTION_VENTING',
    'emergency_stop': 'ACTION_EMERGENCY_STOP',
}

CONFIG_SCHEMA = (
    button.button_schema(HormannButton)
    .extend({
        cv.GenerateID(CONF_HORMANN_HCP1_ID): cv.use_id(HormannHCP1Component),
        cv.Required(CONF_ACTION): cv.enum(ACTION_OPTIONS, lower=True),
    })
    .extend(cv.COMPONENT_SCHEMA)
)


async def to_code(config):
    var = cg.new_Pvariable(config[CONF_ID])
    await cg.register_component(var, config)
    await button.register_button(var, config)
    
    parent = await cg.get_variable(config[CONF_HORMANN_HCP1_ID])
    cg.add(var.set_parent(parent))
    
    action = config[CONF_ACTION]
    cg.add(var.set_action(getattr(hormann_hcp1_ns, ACTION_OPTIONS[action])))

File: components/hormann_hcp1/button.cpp
#include "button.h"
#include "esphome/core/log.h"

namespace esphome {
namespace hormann_hcp1 {

static const char *const TAG = "hormann_hcp1.button";

void HormannButton::dump_config() {
  LOG_BUTTON("", "Hörmann HCP1 Button", this);
  const char *action_str;
  switch (this->action_) {
    case ACTION_IMPULSE: action_str = "Impulse"; break;
    case ACTION_VENTING: action_str = "Venting"; break;
    case ACTION_EMERGENCY_STOP: action_str = "Emergency Stop"; break;
    default: action_str = "Unknown"; break;
  }
  ESP_LOGCONFIG(TAG, "  Action: %s", action_str);
}

void HormannButton::press_action() {
  if (this->parent_ != nullptr && this->action_ != ACTION_NONE) {
    ESP_LOGD(TAG, "Button pressed, triggering action");
    this->parent_->trigger_action(this->action_);
  }
}

}  // namespace hormann_hcp1
}  // namespace esphome
